File: backend/app/domain/websocket_execution.py
from __future__ import annotations

import json
import re
from dataclasses import dataclass
from datetime import datetime
from typing import Any, Literal
from urllib.parse import urlsplit, urlunsplit

WebSocketExecutionStatus = Literal[
    "pending", "queued", "running", "passed", "failed", "error", "cancelled", "timeout"
]
WebSocketEventLevel = Literal["info", "warning", "error"]
WebSocketAssertionKind = Literal["encoding", "text_equals", "text_contains", "json_path_equals"]
WEBSOCKET_JSON_PATH_PATTERN = re.compile(r"^\$(?:\.(?:[A-Za-z_][A-Za-z0-9_-]*|\d+))+$")
TERMINAL_WEBSOCKET_EXECUTION_STATUSES: frozenset[WebSocketExecutionStatus] = frozenset(
    {"passed", "failed", "error", "cancelled", "timeout"}
)


def build_websocket_url(base_url: str, path: str) -> str:
    base = urlsplit(base_url)
    request = urlsplit(path)
    if (
        base.scheme not in {"http", "https"}
        or not base.hostname
        or base.username is not None
        or base.password is not None
        or base.query
        or base.fragment
        or request.scheme
        or request.netloc
        or not request.path.startswith("/")
        or request.path.startswith("//")
        or request.fragment
        or "\r" in path
        or "\n" in path
    ):
        raise ValueError("unsafe WebSocket URL")
    try:
        _ = base.port
    except ValueError as exception:
        raise ValueError("invalid WebSocket target port") from exception
    scheme = "wss" if base.scheme == "https" else "ws"
    combined_path = f"{base.path.rstrip('/')}{request.path}"
    return urlunsplit((scheme, base.netloc, combined_path, request.query, ""))


@dataclass(frozen=True, slots=True)
class WebSocketExecutionStartInput:
    environment_id: str
    path: str
    headers: dict[str, str]
    message: str
    timeout_seconds: int
    additional_messages: tuple[str, ...] = ()
    receive_count: int = 1
    ping_interval_seconds: int | None = None
    max_reconnect_attempts: int = 0
    assertions: tuple[WebSocketMessageAssertion, ...] = ()


@dataclass(frozen=True, slots=True)
class WebSocketMessage:
    ordinal: int
    message: str
    encoding: Literal["text", "base64"]
    size_bytes: int


@dataclass(frozen=True, slots=True)
class WebSocketMessageAssertion:
    message_index: int
    kind: WebSocketAssertionKind
    path: str | None
    expected: str

    def validate(self) -> WebSocketMessageAssertion:
        if not 0 <= self.message_index < 20 or not self.expected or len(self.expected) > 4000:
            raise ValueError("invalid WebSocket assertion")
        if self.kind == "encoding":
            if self.path is not None or self.expected not in {"text", "base64"}:
                raise ValueError("invalid WebSocket encoding assertion")
        elif self.kind in {"text_equals", "text_contains"}:
            if self.path is not None:
                raise ValueError("invalid WebSocket text assertion")
        elif self.kind == "json_path_equals":
            if self.path is None or WEBSOCKET_JSON_PATH_PATTERN.fullmatch(self.path) is None:
                raise ValueError("invalid WebSocket JSON path assertion")
            try:
                expected = json.loads(self.expected)
            except json.JSONDecodeError as exception:
                raise ValueError("invalid WebSocket JSON assertion") from exception
            if isinstance(expected, (dict, list)):
                raise ValueError("WebSocket JSON assertion expected value must be scalar")
        else:
            raise ValueError("unsupported WebSocket assertion")
        return self


@dataclass(frozen=True, slots=True)
class WebSocketMessageAssertionResult:
    message_index: int
    kind: WebSocketAssertionKind
    path: str | None
    expected: str
    actual: str | None
    passed: bool
    message: str


def _display(value: object) -> str:
    rendered = json.dumps(value, ensure_ascii=False, separators=(",", ":"))
    return rendered if len(rendered) <= 500 else f"{rendered[:497]}..."


def _json_path_value(document: object, path: str) -> tuple[bool, object]:
    current = document
    for segment in path.split(".")[1:]:
        if isinstance(current, dict) and segment in current:
            current = current[segment]
        elif isinstance(current, list) and segment.isdigit() and int(segment) < len(current):
            current = current[int(segment)]
        else:
            return False, None
    return True, current


def evaluate_websocket_assertions(
    assertions: tuple[WebSocketMessageAssertion, ...],
    responses: tuple[WebSocketMessage, ...],
) -> tuple[WebSocketMessageAssertionResult, ...]:
    results: list[WebSocketMessageAssertionResult] = []
    for assertion in assertions:
        assertion.validate()
        response = (
            responses[assertion.message_index] if assertion.message_index < len(responses) else None
        )
        actual: str | None = None
        passed = False
        if response is not None and assertion.kind == "encoding":
            actual = response.encoding
            passed = actual == assertion.expected
        elif response is not None and assertion.kind in {"text_equals", "text_contains"}:
            if response.encoding == "text":
                actual = response.message
                passed = (
                    actual == assertion.expected
                    if assertion.kind == "text_equals"
                    else assertion.expected in actual
                )
        elif response is not None and assertion.kind == "json_path_equals":
            if response.encoding == "text":
                try:
                    document: Any = json.loads(response.message)
                except json.JSONDecodeError:
                    document = None
                found, value = _json_path_value(document, assertion.path or "")
                if found:
                    actual = _display(value)
                    expected_value = json.loads(assertion.expected)
                    passed = value == expected_value and type(value) is type(expected_value)
        results.append(
            WebSocketMessageAssertionResult(
                assertion.message_index,
                assertion.kind,
                assertion.path,
                assertion.expected,
                actual,
                passed,
                "消息断言通过。" if passed else "消息缺失、编码不匹配或内容与预期不一致。",
            )
        )
    return tuple(results)


@dataclass(frozen=True, slots=True)
class WebSocketExecutionEvent:
    id: str
    ordinal: int
    level: WebSocketEventLevel
    code: str
    message: str
    created_at: datetime


@dataclass(frozen=True, slots=True)
class WebSocketExecution:
    id: str
    workspace_id: str
    environment_id: str | None
    environment_name: str
    base_url: str
    path_template: str
    headers_template: dict[str, str]
    variables: dict[str, str]
    secret_names: tuple[str, ...]
    message_template: str
    timeout_seconds: int
    status: WebSocketExecutionStatus
    progress: int
    pid: int | None
    response_message: str | None
    response_encoding: Literal["text", "base64"] | None
    response_size_bytes: int | None
    duration_ms: int | None
    error_code: str | None
    error_message: str | None
    created_at: datetime
    started_at: datetime | None
    finished_at: datetime | None
    events: tuple[WebSocketExecutionEvent, ...]
    additional_message_templates: tuple[str, ...] = ()
    receive_count: int = 1
    ping_interval_seconds: int | None = None
    max_reconnect_attempts: int = 0
    responses: tuple[WebSocketMessage, ...] = ()
    assertions: tuple[WebSocketMessageAssertion, ...] = ()
    assertion_results: tuple[WebSocketMessageAssertionResult, ...] = ()
    attempt_count: int = 1

    @property
    def can_cancel(self) -> bool:
        return self.status not in TERMINAL_WEBSOCKET_EXECUTION_STATUSES


@dataclass(frozen=True, slots=True)
class WebSocketExecutionInput:
    base_url: str
    path_template: str
    headers_template: dict[str, str]
    variables: dict[str, str]
    secret_names: tuple[str, ...]
    message_template: str
    timeout_seconds: int
    additional_message_templates: tuple[str, ...] = ()
    receive_count: int = 1
    ping_interval_seconds: int | None = None
    max_reconnect_attempts: int = 0
    assertions: tuple[WebSocketMessageAssertion, ...] = ()


@dataclass(frozen=True, slots=True)
class WebSocketExecutionResult:
    message: str
    encoding: Literal["text", "base64"]
    size_bytes: int
    duration_ms: int
    responses: tuple[WebSocketMessage, ...] = ()
    assertion_results: tuple[WebSocketMessageAssertionResult, ...] = ()
    attempt_count: int = 1


@dataclass(frozen=True, slots=True)
class WebSocketExecutionTaskRequest:
    run_id: str

File: frontend/src/api/websocket-execution.ts
import { resolveBackendConnection } from "./backend-connection";
import { ApiClient, ApiClientError } from "./client";

export type WebSocketExecutionStatus =
  | "pending"
  | "queued"
  | "running"
  | "passed"
  | "failed"
  | "error"
  | "cancelled"
  | "timeout";

export interface WebSocketExecutionStartInput {
  environment_id: string;
  path: string;
  headers: Record<string, string>;
  message: string;
  timeout_seconds: number;
  additional_messages: string[];
  receive_count: number;
  ping_interval_seconds: number | null;
  max_reconnect_attempts: number;
  assertions: WebSocketMessageAssertion[];
}

export type WebSocketAssertionKind =
  | "encoding"
  | "text_equals"
  | "text_contains"
  | "json_path_equals";

export interface WebSocketMessageAssertion {
  message_index: number;
  kind: WebSocketAssertionKind;
  path: string | null;
  expected: string;
}

export interface WebSocketMessageAssertionResult extends WebSocketMessageAssertion {
  actual: string | null;
  passed: boolean;
  message: string;
}

export interface WebSocketMessage {
  ordinal: number;
  message: string;
  encoding: "text" | "base64";
  size_bytes: number;
}

export interface WebSocketExecutionEvent {
  id: string;
  ordinal: number;
  level: "info" | "warning" | "error";
  code: string;
  message: string;
  created_at: string;
}

export interface WebSocketExecution {
  id: string;
  workspace_id: string;
  environment_id: string | null;
  environment_name: string;
  base_url: string;
  path_template: string;
  headers_template: Record<string, string>;
  message_template: string;
  additional_message_templates: string[];
  receive_count: number;
  ping_interval_seconds: number | null;
  max_reconnect_attempts: number;
  timeout_seconds: number;
  status: WebSocketExecutionStatus;
  progress: number;
  response_message: string | null;
  response_encoding: "text" | "base64" | null;
  response_size_bytes: number | null;
  duration_ms: number | null;
  responses: WebSocketMessage[];
  assertions: WebSocketMessageAssertion[];
  assertion_results: WebSocketMessageAssertionResult[];
  attempt_count: number;
  error_code: string | null;
  error_message: string | null;
  created_at: string;
  started_at: string | null;
  finished_at: string | null;
  events: WebSocketExecutionEvent[];
}

const statuses = new Set<WebSocketExecutionStatus>([
  "pending",
  "queued",
  "running",
  "passed",
  "failed",
  "error",
  "cancelled",
  "timeout",
]);

function nullableString(value: unknown): boolean {
  return value === null || typeof value === "string";
}

function stringMap(value: unknown): value is Record<string, string> {
  return (
    typeof value === "object" &&
    value !== null &&
    !Array.isArray(value) &&
    Object.values(value).every((item) => typeof item === "string")
  );
}

function isEvent(value: unknown): value is WebSocketExecutionEvent {
  if (typeof value !== "object" || value === null) return false;
  const item = value as Partial<WebSocketExecutionEvent>;
  return (
    typeof item.id === "string" &&
    typeof item.ordinal === "number" &&
    ["info", "warning", "error"].includes(item.level ?? "") &&
    typeof item.code === "string" &&
    typeof item.message === "string" &&
    typeof item.created_at === "string"
  );
}

function isAssertion(value: unknown): value is WebSocketMessageAssertion {
  if (typeof value !== "object" || value === null) return false;
  const item = value as Partial<WebSocketMessageAssertion>;
  return (
    typeof item.message_index === "number" &&
    ["encoding", "text_equals", "text_contains", "json_path_equals"].includes(item.kind ?? "") &&
    nullableString(item.path) &&
    typeof item.expected === "string"
  );
}

function isAssertionResult(value: unknown): value is WebSocketMessageAssertionResult {
  if (!isAssertion(value)) return false;
  const item = value as Partial<WebSocketMessageAssertionResult>;
  return nullableString(item.actual) && typeof item.passed === "boolean" && typeof item.message === "string";
}

function isMessage(value: unknown): value is WebSocketMessage {
  if (typeof value !== "object" || value === null) return false;
  const item = value as Partial<WebSocketMessage>;
  return typeof item.ordinal === "number" && typeof item.message === "string" &&
    ["text", "base64"].includes(item.encoding ?? "") && typeof item.size_bytes === "number";
}

function isExecution(value: unknown): value is WebSocketExecution {
  if (typeof value !== "object" || value === null) return false;
  const item = value as Partial<WebSocketExecution>;
  return (
    typeof item.id === "string" &&
    typeof item.workspace_id === "string" &&
    nullableString(item.environment_id) &&
    typeof item.environment_name === "string" &&
    typeof item.base_url === "string" &&
    typeof item.path_template === "string" &&
    stringMap(item.headers_template) &&
    typeof item.message_template === "string" &&
    Array.isArray(item.additional_message_templates) &&
    item.additional_message_templates.every((message) => typeof message === "string") &&
    typeof item.receive_count === "number" &&
    (item.ping_interval_seconds === null || typeof item.ping_interval_seconds === "number") &&
    typeof item.max_reconnect_attempts === "number" &&
    typeof item.timeout_seconds === "number" &&
    statuses.has(item.status as WebSocketExecutionStatus) &&
    typeof item.progress === "number" &&
    nullableString(item.response_message) &&
    (item.response_encoding === null ||
      item.response_encoding === "text" ||
      item.response_encoding === "base64") &&
    (item.response_size_bytes === null || typeof item.response_size_bytes === "number") &&
    (item.duration_ms === null || typeof item.duration_ms === "number") &&
    Array.isArray(item.responses) && item.responses.every(isMessage) &&
    Array.isArray(item.assertions) && item.assertions.every(isAssertion) &&
    Array.isArray(item.assertion_results) && item.assertion_results.every(isAssertionResult) &&
    typeof item.attempt_count === "number" &&
    nullableString(item.error_code) &&
    nullableString(item.error_message) &&
    typeof item.created_at === "string" &&
    nullableString(item.started_at) &&
    nullableString(item.finished_at) &&
    Array.isArray(item.events) &&
    item.events.every(isEvent)
  );
}

function validateExecution(value: unknown): WebSocketExecution {
  if (!isExecution(value)) {
    throw new ApiClientError("后端 WebSocket 执行响应格式不正确。", {
      code: "INVALID_RESPONSE",
    });
  }
  return value;
}

function validateExecutions(value: unknown): WebSocketExecution[] {
  if (!Array.isArray(value) || !value.every(isExecution)) {
    throw new ApiClientError("后端 WebSocket 执行列表格式不正确。", {
      code: "INVALID_RESPONSE",
    });
  }
  return value;
}

async function client(): Promise<ApiClient> {
  const connection = await resolveBackendConnection();
  return new ApiClient(connection.baseUrl, connection.token);
}

export async function startWebSocketExecution(
  workspaceId: string,
  input: WebSocketExecutionStartInput,
): Promise<WebSocketExecution> {
  return validateExecution(
    await (await client()).post<unknown>(
      `/api/workspaces/${workspaceId}/websocket-executions`,
      input,
    ),
  );
}

export async function listWebSocketExecutions(
  workspaceId: string,
): Promise<WebSocketExecution[]> {
  return validateExecutions(
    await (await client()).get<unknown>(`/api/workspaces/${workspaceId}/websocket-executions`),
  );
}

export async function getWebSocketExecution(
  workspaceId: string,
  runId: string,
): Promise<WebSocketExecution> {
  return validateExecution(
    await (await client()).get<unknown>(
      `/api/workspaces/${workspaceId}/websocket-executions/${runId}`,
    ),
  );
}

export async function cancelWebSocketExecution(
  workspaceId: string,
  runId: string,
): Promise<WebSocketExecution> {
  return validateExecution(
    await (await client()).post<unknown>(
      `/api/workspaces/${workspaceId}/websocket-executions/${runId}/cancel`,
    ),
  );
}
